sdk/servicebus/service-bus/test/internal/batchReceiver.spec.ts

```diff
@@ -1218,5 +1208,242 @@ describe("Batching Receiver", () => {
     });
+
+    it("returns messages if drain is in progress (receiveAndDelete)", async function(): Promise<
+      void
+    > {
+      // Create the sender and receiver.
+      await beforeEachTest("receiveAndDelete");
+
+      // Send a message so we can be sure when the receiver is open and active.
+      await sender.sendMessages(TestMessage.getSessionSample());
+
+      const messages1 = await receiver.receiveMessages(1, { maxWaitTimeInMs: 5000 });
+      should.equal(
+        messages1.length,
+        1,
+        "Unexpected number of received messages(before disconnect)."
+      );
+
+      const receiverContext = (receiver as ServiceBusSessionReceiverImpl)["_context"];
+      const batchingReceiver = (receiver as ServiceBusSessionReceiverImpl)["_messageSession"];
+
+      // Send a message so we have something to receive.
+      await sender.sendMessages(TestMessage.getSessionSample());
+
+      // We want to simulate a disconnect once the batching receiver is draining.
+      // We can detect when the receiver enters a draining state when `addCredit` is
+      // called while `drain` is set to true.
+      let didRequestDrain = false;
+      const addCredit = batchingReceiver["link"]!.addCredit;
+      batchingReceiver["link"]!.addCredit = function(credits) {
+        // This makes sure the receiveMessages doesn't end because of draining before the disconnect is triggered
+        // Meaning.. the "resolving the messages" can only happen through the onDetached triggered by disconnect
+        batchingReceiver["link"]!.removeAllListeners(ReceiverEvents.receiverDrained);
+        addCredit.call(this, credits);
+        if (batchingReceiver["link"]!.drain) {
+          didRequestDrain = true;
+          // Simulate a disconnect being called with a non-retryable error.
+          receiverContext.connection["_connection"].idle();
+        }
+      };
+
+      // Purposefully request more messages than what's available
+      // so that the receiver will have to drain.
+      const messages2 = await receiver.receiveMessages(10, { maxWaitTimeInMs: 5000 });
+
+      didRequestDrain.should.equal(true, "Drain was not requested.");
+      messages2.length.should.equal(
+        1,
+        "Unexpected number of messages received(during disconnect)."
+      );
+
+      await sender.sendMessages(TestMessage.getSessionSample());
+
+      try {
+        // New receiveMessages should fail because the session lock would be lost due to the disconnection
+        await receiver.receiveMessages(1, { maxWaitTimeInMs: 5000 });
+        assert.fail("Receive messages call should have failed since the lock was lost");
+      } catch (error) {
+        should.equal(
+          error.message,
+          `The session lock has expired on the session with id ${TestMessage.sessionId}`,
+          "Unexpected error thrown"
+        );
+        // wait for the 2nd message to be received.
+        await receiver.close();
```

Comment:
Now technically we don't need this close as close should have already been called by the time we're seeing this exception getting thrown, correct?

Reply:
removed